### PR TITLE
Frozen InMemoryLayers & checkpointing without holding lock

### PR DESCRIPTION
**Pull request by patins**
I've taken the advice from #572 and avoided creating FrozenLayer.

I ran some benchmarks on my system and noticed that I was still seeing some downward TPS spikes every 5 seconds, so maybe this isn't quite solved yet. I'll have to dig in more there, but my first guess that we're copying lots of data or logging a ton.

Haven't figured out how to have multiple frozen layers for the drop solution mentioned by @lubennikovaav, but we should think more about that now

### CODE REVIEW

File: pageserver/src/layered_repository.rs [hunk 1]
```diff
@@ -1272,13 +1273,30 @@ impl LayeredTimeline {
...
+
+            drop(layers);
+            let new_historics = frozen.write_to_disk(self)?;
+            layers = self.layers.lock().unwrap();
+
+            layers.remove_historic(frozen.as_ref());
+
+            if let Some(last_historic) = new_historics.last() {
+                if let Some(new_open) = &maybe_new_open {
+                    new_open.update_predecessor(Arc::clone(last_historic));
```
Comment (patins) [inline pageserver/src/layered_repository.rs:1296]: TODO: Add back assert checking old predecessor

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 2]
```diff
@@ -129,6 +129,11 @@ impl Layer for InMemoryLayer {
     }
 
     fn get_end_lsn(&self) -> Lsn {
+        if let Some(end_lsn) = self.end_lsn {
+            // TODO is this correct?
```
Comment (patins) [inline pageserver/src/layered_repository/inmemory_layer.rs:133]: @hlinnaka is this appropriate?

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 3]
```diff
@@ -62,9 +60,11 @@ pub struct InMemoryLayerInner {
     ///
     segsizes: BTreeMap<Lsn, u32>,
 
-    /// True if freeze() has been called on the layer, indicating it no longer
-    /// accepts writes.
-    frozen: bool,
+    /// This layer is in the process of being replaced, no writes are allowed.
+    writeable: bool,
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository/inmemory_layer.rs:64]: I suggest rewording the comment so that it matches the variable name.

This variable should help to implement multiple `frozen` layers:
1. When some layer is dropped - mark it as non-writeable.
2. When searching for a layer to write, get the latest open. If it is non-writeable, just create a new open layer.
I think that should work.

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 2 above)
Comment (lubennikovaav) [reply pageserver/src/layered_repository/inmemory_layer.rs:133]: I think we should just return the `end_lsn` as is and implement bound "exclusiveness" on a caller's side by using strict comparison operators where necessary.

File: pageserver/src/layered_repository/inmemory_layer.rs [hunk 4]
```diff
@@ -495,45 +514,31 @@ impl InMemoryLayer {
...
-    pub fn freeze(
-        &self,
-        cutoff_lsn: Lsn,
-        // This is needed just to call materialize_page()
-        timeline: &LayeredTimeline,
-    ) -> Result<SuccessorLayers> {
+    /// Splits `self` into two InMemoryLayers: `frozen` and `open`.
+    /// All data up to and including `cutoff_lsn` (or the drop LSN, if dropped)
+    /// is copied to `frozen`, while the remaining data is copied to `open`.
+    /// After completion, self is non-writeable, but not frozen.
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository/inmemory_layer.rs:532]: Why do we need to copy data to the new open layer? Can't we just freeze the layer as is?

Comment (patins) [reply pageserver/src/layered_repository/inmemory_layer.rs:532]: When data has been written since the cutoff I don't want this data to disappear to other readers by making end_lsn smaller (when writeable it's u64::MAX)

Comment (lubennikovaav) [reply pageserver/src/layered_repository/inmemory_layer.rs:532]: Hmm, I don't get it.
We receive `last_record_lsn` that is used as `cutoff` under lock and then mark the layer as `non-writeable`. How can any concurrent write since happen after cutoff?

**Review by hlinnaka — APPROVED** [review]

File: pageserver/src/layered_repository/inmemory_layer.rs (hunk 3 above)
Comment (patins) [reply pageserver/src/layered_repository/inmemory_layer.rs]: I see. Would the checkpointer thread still be responsible for flushing these non-writeable layers out to disk?
